File: problem/0509_FibonacciNumber.js
// 509_FibonacciNumber.js
// https://leetcode.com/problems/fibonacci-number/
// The Fibonacci numbers, commonly denoted F(n) form a sequence, called the Fibonacci sequence, such that each number is the sum of the two preceding ones, starting from 0 and 1. That is,

// F(0) = 0, F(1) = 1
// F(n) = F(n - 1) + F(n - 2), for n > 1.
// Given n, calculate F(n).

/**
 * @param {number} n
 * @return {number}
 */

// recursion
var fib = function(n) {
    if (n ===0) {
      return 0
    } else if (n === 1){
      return 1
    } else {
      return fib(n-1) + fib(n-2)
    }
};

// DP
var fib = function(n) {
    if(n < 2) return n;
    let a = 0;
    let b = 1;
    for(let i=1; i<n; i++){
        [a, b] = [b, a+b]
    }
    return b
};

console.log(fib(4)) //3

File: c#-solution/0509. Fibonacci Number.cs
// 0509. Fibonacci Number
// https://leetcode.com/problems/fibonacci-number/
// Tag: C#, Array, Dynamic Programming, Easy, fibonacci 

public class Solution
{
  public int Fib(int n)
  {
    if (n <= 1) return n;
    int[] nums = new int[n + 1];
    nums[1] = 1;
    for (int i = 2; i <= n; i++)
    {
      nums[i] = nums[i - 1] + nums[i - 2];
    }
    return nums[n];
  }
}
// TC: O(n)
// SC: O(n)
